### ISSUE TITLE
[RFC] Programmatic payments

### ISSUE THREAD

**Issue by witoszekdev** [saleor/saleor#12843]
# Problem

There is no way in Saleor to tokenize (store), retrieve and manage saved payment methods and then charge customers using those methods (also referred to as programmatic charges or offline flow).

Existing solutions (e.g. `availablePaymentGateways`) work only with Payment plugins (not apps) and don’t have any management API that could be used by customers.

## 🏁 Goals

- Add API for fetching and managing stored Payment Methods in Payment Provider Apps
  - Add mutation inspired by existing `storedPaymentSources` for fetching Payment Methods
  - Add sync webhooks to allow communication for fetching and managing stored payment methods in Payment Apps
  - Add mutations and sync webhooks that work similar to Transaction Flow to tokenize new payment methods
  - Add mutations and sync webhooks for updating and deleting payment methods
  - Add mutation for Payment Provider Apps and notifications when payment method expires / is no longer available
  - Add mutation for setting user’s default Payment Method per channel
- Create API for making programmatic charges by staff members from customer stored payment methods (aka offline flow)
  - We should create a new webhook similar to `transaction-initialize-session` and `transaction-process-session webhook` — that would return information about charge / auth success, failure, manual action required (3d secure) for manual charges
  - Add async webhook for receiving notifications about failed manual charges (so that user can re-authenticate)
- Incurred subscription charges should be stored in Saleor as Transactions

## ❌ Non goals

- Create an API for managing subscription contracts / mandates
  - This will be tackled in a separate RFC in future

> **Note**
> In this RFC webhook responses are written as TypeScript types / interfaces
>
> Please keep in mind that `:` means that the field is required and `?:` means that the field is optional
>
> Disrimination unions look like this:
>
> ```ts
> type Response = {ok: true; data: JSON} | {ok: false; error: string}`
> ```
>
> this means: if we're returning `ok = true`, `data` is also in the response; if we're returning `ok = false` then `error` must be also in the response

## Status

```[tasklist]
- [ ] https://github.com/saleor/saleor/issues/13384
- [ ] https://github.com/saleor/saleor/issues/13395
- [ ] https://github.com/saleor/saleor/issues/13386
- [ ] https://github.com/saleor/saleor/issues/13401
```

# Workflow

## Fetching list of saved payment methods

```mermaid
sequenceDiagram
    actor User
    User->>Storefront: Visits a storefront page for managing saved payment methods
    Storefront->>Saleor: User.storedPaymentMethods query
    par Fetching from Payment Apps
    Saleor->>Adyen App: LIST_PAYMENT_METHODS webhook
    Adyen App->>Saleor: List of stored payment methods for the user
    end
    Saleor->>Storefront: Views list of stored payment methods
    Storefront->>User: Sees a list of saved payment methods
```

## Fetching list of saved payment methods in checkout session

> **Note**
> This flow is already supported by Saleor (doesn't require changes from this RFC)

```mermaid
sequenceDiagram
    actor User
    User->>Storefront: Visits a checkout page
    Storefront->>Saleor: paymentGatewayInitialize mutation
    Saleor->>Adyen App: PAYMENT_GATEWAY_INITIALIZE_SESSION webhook
    Adyen App->>Adyen API: Fetch available payment methods<br /> and user stored payment methods
    Adyen API->>Adyen App: List of stored payment methods
    Adyen App->>Saleor: PaymentGateway data with config
    Saleor->>Storefront: List of PaymentGateway data and configs
    Storefront->>User: Sees a list of available payment methods<br /> and stored paymetn methods
```

## Deleting saved payment method

```mermaid
sequenceDiagram
    actor User
    User ->> Storefront: Delete payment method
    Storefront->>+Saleor: paymentMethodRequestDelete<br>mutation to<br>request deletion
    Saleor->>+PaymentApp: PAYMENT_METHOD_REQUEST_DELETE<br>webhook
    PaymentApp->>+PaymentProviderAPI: Process the delete request
    PaymentProviderAPI-->>-PaymentApp: OK
    PaymentApp-->>-Saleor: Return ok
    Saleor-->>-Storefront: Return ok
    Storefront -->> User: Payment method remvoed
```

## Using saved payment method to pay for the order

```mermaid
sequenceDiagram
    actor User
    User->>Storefront: Visits a checkout page to complete the checkout
    rect rgba(194, 219, 255, 0.2)
        Storefront->>Saleor: Channel.availablePaymentGateways query
        Saleor->>Storefront: Returns list of payment gateways
        Storefront->>Saleor: paymentGatewayInitialize mutation
        par Initlizing payment gateways
        Saleor->>Adyen App: PAYMENT_GATEWAY_INITIALIZE_SESSION webhook
        Adyen App->>Adyen API: Create session
        Adyen API->>Adyen App: 
        Adyen App->>Saleor: Return session data for dropin
        end
        Saleor->>Storefront: Returns session data from payment apps
    end
    opt Optional way of fetching payment methods - in case methods are not returned in session data from payment app
        Storefront->>Saleor: Checkout.storedPaymentMethods
        par Fetching payment methods from Payment Provider apps
            Saleor->>Adyen App: LIST_PAYMENT_METHODS webhook
            Adyen App->>Adyen API: Fetch stored payment methods for provided user in channe's merchant account
            Adyen API->>Adyen App: Return list of stored payment methods
            Adyen App->>Saleor: Respond to webhook call
        end
        Saleor->>Storefront: Returns list of PaymentMethod objects
    end
    Storefront->>User: Sees a list of available payment methods
    User->>Storefront: Selects a saved payment method
    User->>Storefront: Enters CVC for saved credit card
    Storefront->>Saleor: transactionInitialize mutation
    Saleor->>Adyen App: TRANSACTION_INITIALIZE_SESSION webhook
    Adyen App->>Adyen API: Charge customer from tokenized payment method
    Adyen API->>Adyen App: Return pspReference
    Adyen App->>Saleor: Respond to TRANSACTION_INITIALIZE_SESSION webhook
    Saleor->>Storefront: Return created transaction
    note over Adyen API: processes charge asynchronously
    Adyen API->>Adyen App: Charge result + saved credit card token
    Adyen App->>Saleor: transactionEventReport mutation
```

## Storing new payment method - Adyen example

```mermaid
sequenceDiagram
    actor User
    User->>Storefront: Visits a storefront to<br>store the new payment method
        Storefront->>+Saleor: Channel.availablePaymentGateways query
        Saleor-->>-Storefront: Returns list of payment gateways
        Storefront->>+Saleor: paymentGatewayInitializeTokenization<br>mutation to<br>initialize the paymentApp
        Saleor->>Adyen App: PAYMENT_GATEWAY_INITIALIZE_TOKENIZATION<br>webhook
        Adyen App->>+Adyen API: Fetch list of enabled payment methods<br>in merchant account
        Adyen API-->>-Adyen App: List of payment methods
        Adyen App-->>Saleor: Return list of payment methods and client key
        Saleor-->>-Storefront: Returns data from<br>payment apps
        Storefront ->>+ Saleor: paymentMethodInitializeTokenization<br>mutation to check the card<br>and tokenize it
        Saleor ->>+ Adyen App: PAYMENT_METHOD_INITIALIZE_TOKENIZATION<br>webhook
        Adyen App ->>+ Adyen API: BIN check and tokenization
        Adyen API -->>- Adyen App: ID of the saved payment method
        alt No additional actions are required
            Adyen App -->>- Saleor: Result of the tokenization<br>(additionalActionRequired = false)
            Saleor -->>- Storefront: Result of the tokenization
        else Additional action is required (e.g. 3D Secure)
            Adyen App-->>+Saleor: Result of the tokenization<br>(additionalActionRequired = true)
            Saleor -->>- Storefront: Result of the tokenization
            loop Processing in loop. Until finalizing all additional actions
                Storefront ->>+ Saleor: paymentMethodProcessTokenization<br>mutation to process additional action
                Saleor ->>+ Adyen App: PAYMENT_METHOD_PROCESS_TOKENIZATION<br>webhook
                Adyen App ->>+ Adyen API: Process<br> additional<br>action
                Adyen API -->>- Adyen App: Result of<br>additional<br>action
                Adyen App -->>- Saleor: Result of<br>additional<br>action
                Saleor -->>- Storefront: Result of<br>additional<br>action
            end
        end
```

## Storing new payment method - Stripe example

```mermaid
sequenceDiagram
    actor User
    User->>Storefront: Visits a storefront to<br>store the new payment method
    Storefront->>+Saleor: Channel.availablePaymentGateways query
    Saleor-->>-Storefront: Returns list of payment gateways
    Storefront ->>+ Saleor: paymentMethodInitializeTokenization<br>mutation
    Saleor ->>+ Stripe App: PAYMENT_METHOD_INITIALIZE_TOKENIZATION<br>webhook
    Stripe App ->>+ Stripe API: Create SetupIntent
    Stripe API -->>- Stripe App: Return SetupIntent and client secret
    Stripe App -->>- Saleor: Return client secret
    Saleor -->>- Storefront: Return client secret
    User ->> Storefront: Enter credit card data
    Storefront ->>+ Stripe API: Confirm setup
    Stripe API -->>- Storefront: Setup is completed
    opt Additional actions are required
        Storefront ->>+Stripe API: Confirm payment
        Stripe API -->>-Storefront: Payment is confirmed
    end
```

## Charging customer from saved Payment Method in Saleor Dashboard

```mermaid
sequenceDiagram
    actor CS as Customer Support

    CS->>Saleor: paymentMethodCharge mutation
    Saleor->>Adyen App: PAYMENT_METHOD_CHARGE webhook
    Adyen App->>Adyen API: Charge customer with<br> provided paymetn method ID
    Adyen API->>Adyen App: Returns pspReference
    Adyen App->>Saleor: Respond to PAYMENT_METHOD_CHARGE webhook
    note over Adyen App, Saleor: event = CHARGE_REQUEST
    Saleor->>CS: Mutation success
    note over Adyen API: Responds after some time
    alt Charge succeeded
        Adyen API->>Adyen App: Returns success result
        Adyen App->>Saleor: transactionEventReport mutation
        note over Adyen App, Saleor: event = CHARGE_SUCCESS
    else SCA failed
        Adyen API->>Adyen App: Returns failed result
        Adyen App->>Saleor: transactionEventReport mutation
        note over Adyen App, Saleor: event = CHARGE_ACTION_REQUIRED
        Saleor->>Email app: CHARGE_REQUEST_ACTION_REQUIRED webhook
        note over Email app: Webhook call includes actionUrl to redirect user to
        Email app->>Email app: Sends email to user
    end
```

## Next Steps

The following section lists proposed API changes that aren't part of this RFC and will be added in the future

<details>
<summary>Linking `PaymentMethod` and `Transaction` objects</summary>

## Transaction object

If a transaction was created programmatically from a stored payment source, it will be linked in a Transaction object

```graphql
type TransactionSavedPaymentMethod {
  """
  Unique ID of payment method - gatewayId + paymentMethodId
  """
  id: ID!
  """
  Identifier of payment method inside payment gateway
  """
  paymentMethodId: String!
  """
  ID of payment gateway that has this payment source stored
  """
  gatewayId: String!
  """
  Human readable name of payment gateway
  """
  gatewayName: String
}

type Transaction {
  # ...
  paymentMethod: TransactionSavedPaymentMethod
}
```

</details>

<details>
<summary>Storing user's default payment method</summary>

### Get user’s default payment source: `User.defaultPaymentMethod` query

```graphql
type User {
  """
  Returns users default payment method for provided channel
  """
  defaultPaymentMethod(channelId: ID!): PaymentSource
}
```

> **Note**
> This query and mutation will be added later, the API might change a bit

### Update user’s default payment source: `paymentSourceSetDefault` mutation

```graphql
enum PaymentMethodSetDefaultErrorCode {
  GRAPHQL_ERROR
  PAYMENT_SOURCE_NOT_FOUND
  USER_NOT_FOUND
  CHANNEL_NOT_FOUND
}

type PaymentMethodSetDefaultError {
  field: String
  message: String
  code: PaymentMethodSetDefaultErrorCode!
}

type PaymentMethodSetDefault {
  user: User
  paymentSource: PaymentSource
  errors: [PaymentSourceSetDefaultError!]
}

type Mutation {
  """
  Sets payment source as user's default in `User.defaultPaymentMethod`
  """
  paymentMethodSetDefault(
    paymentMethodId: String!
    userId: ID!
    channelId: ID!
  ): PaymentSourceSetDefault
}
```

</details>

<details>
<summary>Notifying Saleor about Payment Methods deleted externally</summary>

## `paymentMethodDeleted` mutation

This mutation is called once a Payment Method is deleted directly through Payment Provider or is no longer available (e.g. it expired / card was blocked by issuing bank)

```graphql
enum PaymentMethodDeleteErrorCode
  GRAPHQL_ERROR
  INVALID
  GATEWAY_ERROR
}

type PaymentMethodDeleteError {
  """
  Name of a field that caused the error. A value of `null` indicates that the error isn't associated with a particular field.
  """
  field: String
  message: String
  code: PaymentMethodDeleteErrorCode!
}

type PaymentMethodDeleteResponse {
  errors: [PaymentMethodDeleteError]
}

type Mutation {
  """
  Reports externally deleted payment method
  """
  paymentMethodDelete(
    """
    ID of the stored payment method
    """
    paymentMethodId: ID!
  ): PaymentMethodDeleteResponse
}
```

## `PAYMENT_METHOD_DELETED` [ASYNC Webhook]

Calling `paymentMethodDelete` causes cache invalidation and `PAYMENT_METHOD_DELETED` webhook to be sent.

```graphql
type PaymentMethodDeleted implements Event {
  """
  Time of the event.
  """
  issuedAt: DateTime!

  """
  Saleor version that triggered the event.
  """
  version: String!

  """
  The user or app that triggered the event.
  """
  issuingPrincipal: IssuingPrincipal!

  """
  The application receiving the webhook.
  """
  recipient: App!

  """
  Payment method which was deleted
  """
  paymentMethod: PaymentMethod!

  """
  User which stored the payment method
  """
  user: User!
}
```

</details>

**Comment by fetimo** [comment]
This looks great and well-considered. I've got a couple of questions:
1. The description for the `SUBSCRIPTION` flow says "Payment method can be used for recurring subcriptuions (fixed amount charged on regular interval)". Would this interval be managed in Saleor? If so, is there any difference between this and `UNSCHEDULED`?
2. Is it down to the client to keep track of subscription orders and run the payment?
3. If a customer doesn't have a card saved should they complete checkout with only subscription items first and non-subscription products separately to generate the correct authorisation?

**Comment by witoszekdev** [comment]
Hi @fetimo 

> The description for the SUBSCRIPTION flow says "Payment method can be used for recurring subcriptuions (fixed amount charged on regular interval)". Would this interval be managed in Saleor? If not, is there any difference between this and UNSCHEDULED?

No, this interval wouldn't be managed by Saleor itself, we're planning to leave that part to app developers, so that the billing schedule fits your business needs.

This distinction is made due to requirements in some payment providers ([like Adyen](https://docs.adyen.com/online-payments/tokenization/create-and-use-tokens?tab=automatic_top_ups_3#create-a-token)) to provide an information whether the charge is initiated:
- by **user** from previously saved payment method 
- by **merchant** on a **scheduled** basis (e.g. monthly subscription)
- by **merchant** on an **unscheduled** basis (e.g. account top-up).

> Is it down to the client to keep track of subscription orders and run the payment?

Yes, as I mentioned, the scheduling of subscriptions and running the payments is out of scope for this RFC.

> If a customer doesn't have a card saved should they complete checkout with only subscription items first and non-subscription products separately to generate the correct authorisation?

Yes, the authorization for subscriptions should be made separately so that the card can be charged later without manual input from the user (e.g. 3D Secure).

E.g. In Adyen, when you save a card with [`recurringProcessingModel`](https://docs.adyen.com/api-explorer/Checkout/latest/post/payments#request-recurringProcessingModel) set to `CardOnFile` (used for one-off, non-subscription payments), the user must be present when making another authorization (usually a CVC of the card and 3D Secure is required).

With `recurrringProcessingModel` set to `Subscription` 3D Secure is only required when making initial authorization.

**Comment by witoszekdev** [comment]
Updated RFC:
- Fixed naming issues in webhooks (paymentSource -> paymentMethod)
- Fixed response examples in `LIST_PAYMENT_METHODS`

Removed from the RFC:

## Transaction object

If a transaction was created programmatically from a stored payment source, it will be linked in a Transaction object

```graphql
type TransactionSavedPaymentMethod {
  """
  Unique ID of payment method - gatewayId + paymentMethodId
  """
  id: ID!
  """
  Identifier of payment method inside payment gateway
  """
  paymentMethodId: String!
  """
  ID of payment gateway that has this payment source stored
  """
  gatewayId: String!
  """
  Human readable name of payment gateway
  """
  gatewayName: String
}

type Transaction {
  # ...
  paymentMethod: TransactionSavedPaymentMethod
}
```

> **Note**
> This feature will be discussed further in the future

## `LIST_PAYMENT_METHOD` response

```tsx
{
  // ...
  supportedChargeFlows: Array<"INTERACTIVE" | "SUBSCRIPTION" | "NON_SCHEDULED">;
}
```

> **Note**
> These additional fields will be added once Subscriptions are added in Saleor

### Get user’s default payment source: `User.defaultPaymentMethod` query

```graphql
type User {
  """
  Returns users default payment method for provided channel
  """
  defaultPaymentMethod(channelId: ID!): PaymentSource
}
```

> **Note**
> This query and mutation will be added later, the API might change a bit

### Update user’s default payment source: `paymentSourceSetDefault` mutation

```graphql
enum PaymentMethodSetDefaultErrorCode {
  GRAPHQL_ERROR
  PAYMENT_SOURCE_NOT_FOUND
  USER_NOT_FOUND
  CHANNEL_NOT_FOUND
}

type PaymentMethodSetDefaultError {
  field: String
  message: String
  code: PaymentMethodSetDefaultErrorCode!
}

type PaymentMethodSetDefault {
  user: User
  paymentSource: PaymentSource
  errors: [PaymentSourceSetDefaultError!]
}

type Mutation {
  """
  Sets payment source as user's default in `User.defaultPaymentMethod`
  """
  paymentMethodSetDefault(
    paymentMethodId: String!
    userId: ID!
    channelId: ID!
  ): PaymentSourceSetDefault
}
```

## Notifications about changes in payment sources

> **Note**
> This notification will be removed because currently we've found that none of the payment providers we support sends a notification before the card expires.
> Furthermore the requirement to send such notification to Saleor for app creators based on card expiration date would be too much

### About to expire: `PAYMENT_SOURCE_EXPIRING` [ASYNC webhook]

Notification is sent by Payment Provider app before a payment method expires

#### Mutation

```graphql
enum PaymentSourceAboutToExpireErrorType {
  GRAPHQL_ERROR
  INVALID
}

type PaymentSourceAboutToExpireError {
  field: String
  message: String
  code: PaymentSourceAboutToExpireErrorType!
}

type PaymentSourceAboutToExpire {
  paymentSource: PaymentSource
  errors: [PaymentSourceAboutToExpireError!]
}

type Mutation {
  paymentSourceAboutToExpire(
    paymentSourceId: String!
    expiresAt: Date!
  ): PaymentSourceAboutToExpire
}
```

#### Async Webhook request

```graphql
type PaymentSourceExpiring implements Event {
  """
  Time of the event.
  """
  issuedAt: DateTime

  """
  Saleor version that triggered the event.
  """
  version: String

  """
  The user that triggered the event.
  """
  issuingPrincipal: User

  """
  The application receiving the webhook.
  """
  recipient: App

  """
  Payment method the notification was issued for
  """
  sourceObject: PaymentMethod

  """
  Date when the payment source will expire
  """
  expiresAt: Date
}
```
